github.com/cockroachdb/cockroach/pkg/storage: TestReplicateRestartAfterTruncationWithRemoveAndReAdd failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/c93feab2c26214456777ac30ceeb8d23f142b1fa

Parameters:
```
TAGS=
GOFLAGS=-race
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=778249&tab=buildLog

```
=== RUN   TestReplicateRestartAfterTruncationWithRemoveAndReAdd
I180714 05:58:34.484880 61624 storage/store.go:1480  [s1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I180714 05:58:34.485005 61624 gossip/gossip.go:354  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:38411" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > 
W180714 05:58:34.489977 61624 gossip/gossip.go:1316  [n2] no incoming or outgoing connections
I180714 05:58:34.490202 61624 storage/store.go:1480  [s2] [n2,s2]: failed initial metrics computation: [n2,s2]: system config not yet available
I180714 05:58:34.490251 61713 gossip/client.go:129  [n2] started gossip client to 127.0.0.1:38411
I180714 05:58:34.490261 61624 gossip/gossip.go:354  [n2] NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:40965" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > 
W180714 05:58:34.548553 61624 gossip/gossip.go:1316  [n3] no incoming or outgoing connections
I180714 05:58:34.548784 61624 storage/store.go:1480  [s3] [n3,s3]: failed initial metrics computation: [n3,s3]: system config not yet available
I180714 05:58:34.548857 61624 gossip/gossip.go:354  [n3] NodeDescriptor set to node_id:3 address:<network_field:"tcp" address_field:"127.0.0.1:38505" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > 
I180714 05:58:34.548789 61721 gossip/client.go:129  [n3] started gossip client to 127.0.0.1:38411
I180714 05:58:34.553519 61624 storage/replica_raftstorage.go:520  [s1,r1/1:/M{in-ax}] generated preemptive snapshot ddd961d2 at index 15
I180714 05:58:34.553574 61624 rpc/nodedialer/nodedialer.go:79  [s1,r1/1:/M{in-ax}] connection to n2 established
I180714 05:58:34.554467 61624 storage/store_snapshot.go:605  [s1,r1/1:/M{in-ax}] streamed snapshot to (n2,s2):?: kv pairs: 47, log entries: 5, rate-limit: 2.0 MiB/sec, 1ms
I180714 05:58:34.554829 61993 storage/replica_raftstorage.go:737  [s2,r1/?:{-}] applying preemptive snapshot at index 15 (id=ddd961d2, encoded size=8159, 1 rocksdb batches, 5 log entries)
I180714 05:58:34.555838 61993 storage/replica_raftstorage.go:743  [s2,r1/?:/M{in-ax}] applied preemptive snapshot in 1ms [clear=0ms batch=0ms entries=0ms commit=1ms]
I180714 05:58:34.556252 61624 storage/replica_command.go:743  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:/M{in-ax} [(n1,s1):1, next=2]
I180714 05:58:34.557448 61624 storage/replica.go:3596  [txn=89c6bb68,s1,r1/1:/M{in-ax}] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I180714 05:58:34.558207 61624 storage/replica_raftstorage.go:520  [s1,r1/1:/M{in-ax}] generated preemptive snapshot 075f8de7 at index 17
I180714 05:58:34.558255 61624 rpc/nodedialer/nodedialer.go:79  [s1,r1/1:/M{in-ax}] connection to n3 established
I180714 05:58:34.558906 61624 storage/store_snapshot.go:605  [s1,r1/1:/M{in-ax}] streamed snapshot to (n3,s3):?: kv pairs: 50, log entries: 7, rate-limit: 2.0 MiB/sec, 1ms
I180714 05:58:34.559196 61962 storage/replica_raftstorage.go:737  [s3,r1/?:{-}] applying preemptive snapshot at index 17 (id=075f8de7, encoded size=9089, 1 rocksdb batches, 7 log entries)
I180714 05:58:34.560624 61962 storage/replica_raftstorage.go:743  [s3,r1/?:/M{in-ax}] applied preemptive snapshot in 1ms [clear=0ms batch=0ms entries=0ms commit=1ms]
I180714 05:58:34.561052 61624 storage/replica_command.go:743  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, next=3]
I180714 05:58:34.564316 61909 rpc/nodedialer/nodedialer.go:79  connection to n1 established
I180714 05:58:34.565991 61624 storage/replica.go:3596  [txn=2de70651,s1,r1/1:/M{in-ax}] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I180714 05:58:34.705273 61624 storage/replica_command.go:743  [s1,r1/1:/M{in-ax}] change replicas (REMOVE_REPLICA (n2,s2):2): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4]
I180714 05:58:34.707244 61624 storage/replica.go:3596  [txn=3a657857,s1,r1/1:/M{in-ax}] proposing REMOVE_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n3,s3):3] next=4
I180714 05:58:34.712048 62117 storage/store.go:3567  [s2,r1/2:/M{in-ax}] added to replica GC queue (peer suggestion)
I180714 05:58:34.712074 62125 storage/store.go:3567  [s2,r1/2:/M{in-ax}] added to replica GC queue (peer suggestion)
W180714 05:58:34.712157 62216 storage/replica.go:6075  [s2,r1/2:/M{in-ax}] could not acquire lease for range gossip: r1 was not found
W180714 05:58:34.712172 62216 storage/store.go:1576  [s2,r1/2:/M{in-ax}] could not gossip node liveness: r1 was not found
I180714 05:58:34.713014 61624 storage/store.go:2513  [replicaGC,s2,r1/2:/M{in-ax}] removing replica
I180714 05:58:34.713251 61624 storage/replica.go:835  [replicaGC,s2,r1/2:/M{in-ax}] removed 47 (42+5) keys in 0ms [clear=0ms commit=0ms]
I180714 05:58:34.713747 61624 storage/replica_raftstorage.go:520  [s1,r1/1:/M{in-ax}] generated preemptive snapshot 6e41897e at index 25
I180714 05:58:34.713863 61624 rpc/nodedialer/nodedialer.go:79  [s1,r1/1:/M{in-ax}] connection to n2 established
I180714 05:58:34.714478 61624 storage/store_snapshot.go:605  [s1,r1/1:/M{in-ax}] streamed snapshot to (n2,s2):?: kv pairs: 58, log entries: 3, rate-limit: 2.0 MiB/sec, 1ms
I180714 05:58:34.714800 62247 storage/replica_raftstorage.go:737  [s2,r1/?:{-}] applying preemptive snapshot at index 25 (id=6e41897e, encoded size=8571, 1 rocksdb batches, 3 log entries)
I180714 05:58:34.715283 62247 storage/replica_raftstorage.go:743  [s2,r1/?:/M{in-ax}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I180714 05:58:34.715674 61624 storage/replica_command.go:743  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n2,s2):4): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n3,s3):3, next=4]
I180714 05:58:34.717476 61624 storage/replica.go:3596  [txn=1a546085,s1,r1/1:/M{in-ax}] proposing ADD_REPLICA((n2,s2):4): updated=[(n1,s1):1 (n3,s3):3 (n2,s2):4] next=5
I180714 05:58:34.735468 61953 storage/store.go:3586  [s1,r1/1:/M{in-ax}] added to replica GC queue (contacted deleted peer)
I180714 05:58:49.713767 62219 storage/compactor/compactor.go:325  [s2,compactor] purging suggested compaction for range /Min - /Max that contains live data
I180714 05:59:05.144317 61602 gossip/gossip.go:1330  [n3] node has connected to cluster via gossip
I180714 05:59:08.170904 61765 gossip/gossip.go:1330  [n1] node has connected to cluster via gossip
I180714 05:59:11.912417 61744 gossip/gossip.go:1330  [n2] node has connected to cluster via gossip
--- FAIL: TestReplicateRestartAfterTruncationWithRemoveAndReAdd (45.32s)
	client_raft_test.go:1617: r1 was not found

```

This one repros

```
$ make stress PKG=./pkg/storage/ TESTS=TestReplicateRestartAfterTruncationWithRemoveAndReAdd
...
--- FAIL: TestReplicateRestartAfterTruncationWithRemoveAndReAdd (45.33s)
        client_raft_test.go:1617: r1 was not found
FAIL


ERROR: exit status 1

1645 runs completed, 1 failures, over 45s
context canceled
FAIL
FAIL    github.com/cockroachdb/cockroach/pkg/storage    45.487s
Makefile:810: recipe for target 'stress' failed
make: *** [stress] Error 1
```

--- a/pkg/storage/store.go
+++ b/pkg/storage/store.go
@@ -3795,7 +3795,24 @@ func (s *Store) raftTickLoop(ctx context.Context) {
 			rangeIDs = rangeIDs[:0]
 			// Update the liveness map.
 			if s.cfg.NodeLiveness != nil {
-				s.livenessMap.Store(s.cfg.NodeLiveness.GetIsLiveMap())
+				nextMap := s.cfg.NodeLiveness.GetIsLiveMap()
+				for nodeID, isLive := range nextMap {
+					if isLive {
+						continue
+					}
+					// Liveness claims that this node is down, but ConnHealth gets the last say
+					// because we'd rather quiesce a range too little than one too often.
+					//
+					// NB: This has false negatives. If a node doesn't have a conn open to it
+					// when ConnHealth is called, then ConnHealth will return
+					// rpc.ErrNotHeartbeated regardless of whether the node is up or not. That
+					// said, for the nodes that matter, we're likely talking to them via the
+					// Raft transport, so ConnHealth should usually indicate a real problem if
+					// it gives us an error back. The check can also have false positives if the
+					// node goes down after populating the map, but that matters even less.
+					nextMap[nodeID] = (s.cfg.NodeDialer.ConnHealth(nodeID) == nil)
+				}
+				s.livenessMap.Store(nextMap)
 			}
 
 			s.unquiescedReplicas.Lock()

